File: vidRec/CommHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using Newtonsoft.Json.Linq;

namespace vidRec
{
    class CommHandler
    {

        private IConnection connection = null;
        private IModel SendChannel = null;
        private IModel ReceiveChannel = null;

        public CommHandler(string ip)
        {
            var factory = new ConnectionFactory();
            factory.UserName = "audio";
            factory.Password = "audio";
            factory.VirtualHost = "/";
            factory.Port = 5672;
            factory.HostName = ip;
            factory.RequestedHeartbeat = 10;
            factory.AutomaticRecoveryEnabled = true;
            connection = factory.CreateConnection();
        }

        public void send(string exchange, string routing, string content)
        {
            try
            {
                if ( SendChannel==null || SendChannel.IsClosed)
                {
                    SendChannel = connection.CreateModel();
                    if (exchange.Contains("fanout"))
                        SendChannel.ExchangeDeclare(exchange: exchange, type: "fanout", durable: true);
                    else if (exchange.Contains("direct"))
                        SendChannel.ExchangeDeclare(exchange: exchange, type: "direct", durable: true);
                    else if (exchange.Contains("topic"))
                        SendChannel.ExchangeDeclare(exchange: exchange, type: "topic", durable: true);
                    else
                        exchange = "";
                }
                var body = Encoding.UTF8.GetBytes(content);
                SendChannel.BasicPublish(exchange: exchange,
                                            routingKey: routing,
                                            basicProperties: null,
                                            body: body);
            }
            catch (Exception ex)
            {
                Console.WriteLine("CommHandler.send(): " + ex.Message);
            }
        }


        public void listen(string exchange, string routing, Action<string> callback)
        {
            if (ReceiveChannel == null || ReceiveChannel.IsClosed)
            {
                ReceiveChannel = connection.CreateModel();
                if (exchange.Contains("fanout"))
                    ReceiveChannel.ExchangeDeclare(exchange: exchange, type: "fanout", durable: true);
                else if (exchange.Contains("direct"))
                    ReceiveChannel.ExchangeDeclare(exchange: exchange, type: "direct", durable: true);
                else if (exchange.Contains("topic"))
                    ReceiveChannel.ExchangeDeclare(exchange: exchange, type: "topic", durable: true);
                else
                    exchange = "";
            }
            // queue
            var queueName = ReceiveChannel.QueueDeclare().QueueName;
            // binding
            if ("" == exchange)
                Console.WriteLine("[*] The queue binds to the default exchange");
            else
                ReceiveChannel.QueueBind(queue: queueName,
                                         exchange: exchange,
                                         routingKey: routing);
                Console.WriteLine(String.Format("[*] Listening to <{0}> on <{1}>", routing, queueName));
                var consumer = new EventingBasicConsumer(ReceiveChannel);
                consumer.Received += (model, ea) =>
                {
                    var body = ea.Body;
                    string message = Encoding.UTF8.GetString(body);
                    callback(message);
                };
                ReceiveChannel.BasicConsume(queue: queueName,
                                            autoAck: true,
                                            consumer: consumer);
        }


        public void close()
        {
            try {
                if (SendChannel != null && SendChannel.IsOpen)
                    SendChannel.Close();
                if (ReceiveChannel != null && ReceiveChannel.IsOpen)
                    ReceiveChannel.Close();
                connection.Close();
                connection = null;
            }
            catch (Exception ex)
            {
                Console.WriteLine("CommHandler.close(): " + ex.Message);
            }
        }

    }
}

File: vidRec/commHandler.py
# -*- coding: utf-8 -*-

import pika
import threading
import os.path
import os
import sys


LOCALHOST = 'localhost'
#SERVER_IP = '129.161.106.38'
#SERVER_IP = '128.113.21.81' # studio 2
SERVER_IP = '129.161.106.33'
# QUEUES
# SpatialContext.registry
# SpatialContext.command.boss
# SpatialContext.command.worker






class commHandler:
    
    def __init__(self, ip=None):
        if ip is None:
            ip = SERVER_IP
#        credentials = pika.PlainCredentials('cel', 'c0gnitive')
#        self.parameters = pika.ConnectionParameters(ip, 5672, 'cel', credentials, heartbeat_interval=0)
        credentials = pika.PlainCredentials('sc', 'sc')
        self.parameters = pika.ConnectionParameters(ip, 5672, '/', credentials, heartbeat=60000)
        self.connection = pika.BlockingConnection(self.parameters)
        self.SendChannel = None
        self.ReceiveChannel = None
        
        
    def __del__(self):
        if self.SendChannel is not None:
            self.SendChannel.close()
        if self.ReceiveChannel is not None:
            self.ReceiveChannel.close()
        self.connection.close()
    
    
    def send(self, exchange, routing, content):
        # exchange
        if self.SendChannel is None or self.SendChannel.is_closed:
            self.SendChannel = self.connection.channel()
            if 'fanout' in exchange:
                self.SendChannel.exchange_declare(exchange=exchange, exchange_type='fanout', durable=True)
            elif 'direct' in exchange:
                self.SendChannel.exchange_declare(exchange=exchange, exchange_type='direct', durable=True)
            elif 'topic' in exchange:
                self.SendChannel.exchange_declare(exchange=exchange, exchange_type='topic', durable=True)  
            else:
                exchange = ''
        # binding
        try:
            self.SendChannel.basic_publish(exchange=exchange,
                                          routing_key=routing,
                                          body=content)
        except pika.exceptions.ConnectionClosed:
            print(' [*] The previous connection is closed.')
            self.connection = pika.BlockingConnection(self.parameters)
            self.SendChannel = self.connection.channel()
            self.send(exchange, routing, content)


    def __consumer__(self, exchange, routing, callback):
        # exchange
        if 'fanout' in exchange:
            self.ReceiveChannel.exchange_declare(exchange=exchange, exchange_type='fanout', durable=True)
        elif 'direct' in exchange:
            self.ReceiveChannel.exchange_declare(exchange=exchange, exchange_type='direct', durable=True)
        elif 'topic' in exchange:
            self.ReceiveChannel.exchange_declare(exchange=exchange, exchange_type='topic', durable=True)  
        else:
            exchange = ''
        # queue
        result = self.ReceiveChannel.queue_declare('',exclusive=True)
        queue = result.method.queue
        # binding
        if '' == exchange:
            print(' [*] The queue binds to the default exchange')
        else:
            self.ReceiveChannel.queue_bind(exchange=exchange, 
                                    queue=queue,
                                    routing_key=routing)
        print(' [*] Listening to <{}> on <{}>'.format(routing, queue) )
#        try:
        self.ReceiveChannel.basic_consume(callback, queue=queue, no_ack=True)  # for pika==0.12.0
        # self.ReceiveChannel.basic_consume(queue=queue, on_message_callback=callback, auto_ack=True)  # for pika==1.0.1
        self.ReceiveChannel.start_consuming()
#        except pika.exceptions.ConnectionClosed:
#            print(' [*] The previous connection is closed.')
#            self.connection = pika.BlockingConnection(self.parameters)
#            self.SendChannel = self.connection.channel()
#            self.ReceiveChannel = self.connection.channel()
#            self.ReceiveChannel.basic_consume(callback, queue=queue, no_ack=True)
#            self.ReceiveChannel.start_consuming()

    
    def listen(self, exchange, routing, callback):
        self.ReceiveChannel = self.connection.channel()
        consumer_thread = threading.Thread(target=self.__consumer__, args=[exchange, routing, callback])
        consumer_thread.start()
        
    
    def close(self):
        self.connection.close()
        self.connection = None


def callback(ch, method, properties, body):
    print('ch -> {}'.format(ch))
    print('method ->', method)
    print(body.decode("utf-8"))


if __name__ == '__main__':
    commToF = commHandler('129.161.106.25')
#    commToF.listen('amq.topic', 'SpatialContext.mic.username_binding', callback)
    commToF.send('amq.topic', 'SpatialContext.mic.username_binding', 'test')
#    commToF.close()




